clippy_lints/src/methods/owned_to_owned.rs

```diff
@@ -0,0 +1,60 @@
+use crate::utils::{match_trait_method, match_type, span_lint_and_sugg};
+use if_chain::if_chain;
+use rustc_errors::Applicability;
+use rustc_hir as hir;
+use rustc_hir::ExprKind;
+use rustc_lint::LateContext;
+use rustc_middle::ty::{TyS,ImplContainer,TraitContainer};
+use rustc_span::symbol::Symbol;
+use rustc_hir::Expr;
+
+use super::OWNED_TO_OWNED;
+use clippy_utils::is_type_diagnostic_item;
+
+// local for now while I'm massively debugging this
+fn match_diagnostic(cx: &LateContext<'_>, expr: &Expr<'_>, diag_item: Symbol) -> bool {
+    let def_id = cx.typeck_results().type_dependent_def_id(expr.hir_id).unwrap();
+    let associated = cx.tcx.opt_associated_item(def_id).and_then(|associated_item| match associated_item.container {
+        TraitContainer(assoc_def_id) => Some(assoc_def_id),
+        ImplContainer(assoc_def_id) => Some(assoc_def_id),
+    });
+    if let Some(assoc_def_id) = associated {
+        // Except this doesn't work for std::path::Path::to_path_buf as this print returns;
+        //  got associated DefId(1:4405 ~ std[e437]::path::{impl#61}) for DefId(1:4413 ~ std[e437]::path::{impl#61}::to_path_buf) // kind : "std::path::Path"
+        // obviously def_path_str realizes this is in std::path::Path, but the assoc_id is "std[e437]::path::{impl#61}"
+        //  how do I get from ::{impl#61} to ::Path (which has my diagnostic item)
+        println!("got associated {:?} for {:?} // kind : {:?}",assoc_def_id,def_id,cx.tcx.def_path_str(assoc_def_id));
+        println!(" --- {:?} {}",cx.tcx.get_diagnostic_item(diag_item), cx.tcx.get_diagnostic_item(diag_item) == Some(assoc_def_id));
```

Comment:
This doesn't seem to work on struct impls:

The ones on struct impls don't work and give me:
```
got associated DefId(1:4405 ~ std[e437]::path::{impl#61}) for DefId(1:4413 ~ std[e437]::path::{impl#61}::to_path_buf) // kind : "std::path::Path"
 --- Some(DefId(1:4403 ~ std[e437]::path::Path)) false
```

While the ones on traits work and get me:
```
got associated DefId(5:424 ~ alloc[88c7]::borrow::ToOwned) for DefId(5:426 ~ alloc[88c7]::borrow::ToOwned::to_owned) // kind : "std::borrow::ToOwned"
 --- Some(DefId(5:424 ~ alloc[88c7]::borrow::ToOwned)) true
```

I need some way to get from the impl def_id to the struct def_id?